File: apps/VecGCD/VecGCD.cpp
#include <NoCL.h>
#include <Rand.h>

// Euclid's algorithm
int gcd(int x, int y) {
  noclPush();
    while (x != y) {
      noclPush();
        if (x > y)
          x = x-y;
        else
          y = y-x;
      noclPop();
    }
  noclPop();
  return x;
}

// Euclid's algorithm on vectors
struct VecGCD : Kernel {
  int len;
  int *a, *b, *result;

  void kernel() {
    for (int i = threadIdx.x; i < len; i += blockDim.x)
      result[i] = gcd(a[i], b[i]);
  }
};

int main()
{
  // Are we in simulation?
  bool isSim = getchar();

  // Vector size for benchmarking
  int N = isSim ? 100 : 100000;

  // Input and output vectors
  simt_aligned int a[N], b[N], result[N];

  // Initialise inputs
  uint32_t seed = 100;
  for (int i = 0; i < N; i++) {
    a[i] = 1 + (rand15(&seed) & 0xff);
    b[i] = 1 + (rand15(&seed) & 0xff);
  }

  // Instantiate kernel
  VecGCD k;

  // Use single block of threads
  k.blockDim.x = SIMTLanes * SIMTWarps;

  // Assign parameters
  k.len = N;
  k.a = a;
  k.b = b;
  k.result = result;

  // Invoke kernel
  noclRunKernelAndDumpStats(&k);

  // Check result
  bool ok = true;
  for (int i = 0; i < N; i++)
    ok = ok && result[i] == gcd(a[i], b[i]);

  // Display result
  puts("Self test: ");
  puts(ok ? "PASSED" : "FAILED");
  putchar('\n');

  return 0;
}

File: rust/examples/vec_gcd.rs
#![no_std]
#![no_main]
#![allow(unused)]

// Crates being used
// =================

use riscv_rt::entry;

extern crate nocl;
use nocl::*;
use nocl::rand::*;
use nocl::prims::*;

extern crate alloc;
use alloc::vec;
use alloc::vec::*;
use alloc::boxed::*;

// Benchmark
// =========

// Euclid's algorithm
fn gcd(x_arg : u32, y_arg : u32) -> u32 {
  let mut x = x_arg;
  let mut y = y_arg;
  nocl_push();
  while x != y {
    nocl_push();
    if (x > y) {
      x = x-y
    }
    else {
      y = y-x
    }
    nocl_pop();
  }
  nocl_pop();
  return x;
}

struct VecGCD {
  len    : usize,
  a      : Buffer<u32>,
  b      : Buffer<u32>,
  result : Buffer<u32>
}

impl Code for VecGCD {
  #[inline(always)]
  fn run (my : &My, shared : &mut Scratch, params: &mut VecGCD) {
    for i in (my.thread_idx.x .. params.len).step_by(my.block_dim.x) {
      params.result[i] = gcd(params.a[i], params.b[i])
    }
  }
}

#[entry]
fn main() -> ! {
  nocl_init();

  // Vector size for benchmarking
  #[cfg(not(feature = "large_data_set"))]
  const N : usize = 100;
  #[cfg(feature = "large_data_set")]
  const N : usize = 100000;

  // Input and output vectors
  let mut a : Vec<u32> = vec![0; N];
  let mut b : Vec<u32> = vec![0; N];
  let mut result : Vec<u32> = vec![0; N];

  // Initialise inputs
  let mut seed : u32 = 100;
  for i in 0..N {
    a[i] = 1 + (rand15(&mut seed) & 0xff);
    b[i] = 1 + (rand15(&mut seed) & 0xff);
  }

  // Use a single block of threads
  let dims =
    Dims {
      block_dim: Dim2 { x: prims::config::SIMT_WARPS *
                           prims::config::SIMT_LANES,
                        y: 1 },
      grid_dim: Dim2 { x: 1, y: 1 }
    };
 
  // Kernel parameters
  let mut params =
    VecGCD { len    : N,
             a      : a.into(),
             b      : b.into(),
             result : result.into() };

  // Invoke kernel
  let params = nocl_run_kernel_verbose(dims, params);

  // Check result
  let mut ok = true;
  for i in 0 .. N {
    ok = ok && params.result[i] == gcd(params.a[i], params.b[i])
  }

  // Display result
  putstr("Self test: ");
  putstr(if ok { "PASSED" } else { "FAILED" });
  putchar(b'\n');

  // The end
  end()
}
